src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,951 @@
+                }
+                else
+                {
+                    argumentListSyntax = syntaxFacts.GetArgumentListOfInvocationExpression(invocation);
+                }
+
+                editor.ReplaceNode(argumentListSyntax, (currentArgumentListSyntax, _) =>
+                {
+                    var updatedInvocationArguments = syntaxFacts.GetArgumentsOfArgumentList(currentArgumentListSyntax);
+                    var updatedExpression = CreateNewArgumentExpression(expressionEditor,
+                        syntaxFacts, mappingDictionary, parameterToArgumentMap, updatedInvocationArguments);
+                    var allArguments = AddArgumentToArgumentList(updatedInvocationArguments, generator,
+                        updatedExpression.WithAdditionalAnnotations(Formatter.Annotation), insertionIndex);
+                    return syntaxFacts.UpdateArgumentListSyntax(currentArgumentListSyntax, allArguments);
+                });
+            }
+
+            // If you are at the original document, then also introduce the new method and introduce the parameter.
+            if (document.Id == originalDocument.Id)
+            {
+                await UpdateExpressionInOriginalFunctionAsync(originalDocument, expression, containingMethod,
+                    parameterName, editor, allOccurrences, cancellationToken).ConfigureAwait(false);
+                var parameterType = invocationSemanticModel.GetTypeInfo(expression, cancellationToken).ConvertedType ??
+                    invocationSemanticModel.Compilation.ObjectType;
+                var parameter = generator.ParameterDeclaration(name: parameterName,
+                    type: generator.TypeExpression(parameterType));
+                editor.InsertParameter(containingMethod, insertionIndex, parameter);
+            }
+
+            return editor.GetChangedRoot();
+        }
+
+        /// <summary>
+        /// This method iterates through the variables in the expression and maps the variables back to the parameter
+        /// it is associated with. It then maps the parameter back to the argument at the invocation site and gets the
+        /// index to retrieve the updated arguments at the invocation.
+        /// </summary>
+        private TExpressionSyntax CreateNewArgumentExpression(SyntaxEditor editor, ISyntaxFactsService syntaxFacts,
+            Dictionary<TIdentifierNameSyntax, IParameterSymbol> mappingDictionary,
+            ImmutableDictionary<IParameterSymbol, int> parameterToArgumentMap,
+            SeparatedSyntaxList<SyntaxNode> updatedInvocationArguments)
+        {
+            foreach (var (variable, mappedParameter) in mappingDictionary)
+            {
+                var parameterMapped = false;
+                if (parameterToArgumentMap.TryGetValue(mappedParameter, out var index))
+                {
+                    var updatedInvocationArgument = updatedInvocationArguments.ToArray()[index];
```

Comment:
```suggestion
                    var updatedInvocationArgument = updatedInvocationArguments[index];
```

Reply:
stop doign O(n) copies to access single elements :)

Reply:
sorry I mentioned in a comment above I didn't know separatedsyntaxlists accessed elements like arrays and was confused why they didn't have a ValueAt or Get method